src/actions/collapseContext.ts

```diff
@@ -59,8 +60,38 @@ const collapseContext = (state: State, { at }: Options) => {
   }
 
   const thought = getThoughtById(state, head(simplePath))
+
+  // Determine whether the collapsed context can be inserted continuously after the meta attributes
+  // or whether we need to split them and insert separetely.
+  // We need to calculate this upfront since we can't rely on `getRankBefore` and `getRankAfter`
+  // without applying changes to state incrementally.
```

Comment:
But we can apply changes to state incrementally! I think that would be a cleaner approach. 

Rather than doing all these calculations up front, just calculate the insertion point as needed for each child that is being moved. If it is a meta attribute, insert it after the existing meta attributes. Otherwise insert it as normal.

Reply:
Yes, exactly. For solutions with comparable complexity I'll usually choose the more computationally efficient option. But if the efficient solution adds complexity and the function is not highly performance-sensitive, I'll usually choose the solution with less code or that maps more closely to the existing code's semantics, even if it is less efficient.